src/main/scala/io/qbeast/spark/index/OTreeDataAnalyzer.scala

```diff
@@ -0,0 +1,227 @@
+/*
+ * Copyright 2021 Qbeast Analytics, S.L.
+ */
+package io.qbeast.spark.index
+
+import com.typesafe.config.ConfigFactory
+import io.qbeast.core.model._
+import io.qbeast.spark.index.QbeastColumns.{cubeToReplicateColumnName, weightColumnName}
+import io.qbeast.spark.internal.QbeastFunctions.qbeastHash
+import org.apache.spark.sql.expressions.UserDefinedFunction
+import org.apache.spark.sql.functions.{col, udaf}
+import org.apache.spark.sql.{DataFrame, Dataset, SparkSession}
+
+trait OTreeDataAnalyzer {
+
+  /**
+   * Analyze the data to process
+   * @param data the data to index
+   * @param indexStatus the current status of the index
+   * @param isReplication either we are replicating the elements or not
+   * @return the changes to the index
+   */
+  def analyze(
+      data: DataFrame,
+      indexStatus: IndexStatus,
+      isReplication: Boolean): (DataFrame, TableChanges)
+
+}
+
+object DoublePassOTreeDataAnalyzer extends OTreeDataAnalyzer with Serializable {
+
+  /**
+   * Estimates MaxWeight on DataFrame
+   */
+  private val maxWeightEstimation: UserDefinedFunction = udaf(MaxWeightEstimation)
+
+  /**
+   * The minimum cube size per partition registered in configuration
+   */
+  private val minPartitionCubeSize: Int =
+    ConfigFactory.load().getInt("qbeast.index.minPartitionCubeSize")
+
+  private lazy val logger = org.apache.log4j.LogManager.getLogger(this.getClass)
+
+  private[index] def calculateRevisionChanges(
+      data: DataFrame,
+      revision: Revision): Option[RevisionChange] = {
+    val columnStats = revision.columnTransformers.map(_.stats)
+    val columnsExpr = columnStats.flatMap(_.columns)
+    // This is a actions that will be executed on the dataframe
+    val rows = data.selectExpr(columnsExpr: _*).collect()
```

Comment:
Is there a reason for this collect? From the looks of it we only use the first row, and it eats a lot of memory on the driver.

Reply:
it should be a single row, because we are using aggregation

Reply:
Updated to check if selectExpr is empty in 9e5192e